### PR TITLE
state: refactor Open to take OpenParams

### PR DESCRIPTION
**Pull request by axw**
Introduce a state.OpenParams struct, and
update state.Open to use it. Add a Clock
parameter while we're at it.

A followup will introduce an observer for
recording mgo/txn-related metrics.

### CODE REVIEW

File: agent/agentbootstrap/bootstrap_test.go [hunk 1]
```diff
@@ -298,7 +299,13 @@ LXC_BRIDGE="ignored"`[1:])
 	info, ok := cfg.MongoInfo()
 	c.Assert(ok, jc.IsTrue)
 	c.Assert(info.Password, gc.Not(gc.Equals), testing.DefaultMongoPassword)
-	st1, err := state.Open(newCfg.Model(), newCfg.Controller(), info, mongotest.DialOpts(), nil)
+	st1, err := state.Open(state.OpenParams{
```
Comment (anastasiamac) [inline agent/agentbootstrap/bootstrap_test.go]: \o/ So much more understandable/readable with structs :D

File: state/open.go [hunk 2]
```diff
@@ -26,35 +26,76 @@ import (
...
+	// ControllerModelTag is the tag of the controller model.
+	ControllerModelTag names.ModelTag
+
+	// MongoInfo is the mongo.MongoInfo used for dialling the
+	// Mongo connection.
+	MongoInfo *mongo.MongoInfo
+
+	// MongoDialOpts is the mongo.DialOpts used to control how
+	// Mongo connections are made.
+	MongoDialOpts mongo.DialOpts
```
Comment (anastasiamac) [inline state/open.go]: Please either comment what would happen if dialopts are nil or add to validation to ensure that it does not happen.

File: state/open.go [hunk 3]
```diff
@@ -26,35 +26,76 @@ import (
...
 // Open returns unauthorizedError if access is unauthorized.
-func Open(
-	controllerModelTag names.ModelTag,
-	controllerTag names.ControllerTag,
-	info *mongo.MongoInfo, opts mongo.DialOpts,
-	newPolicy NewPolicyFunc,
-) (*State, error) {
-	st, err := open(controllerModelTag, info, opts, newPolicy, clock.WallClock)
+func Open(args OpenParams) (*State, error) {
+	if err := args.Validate(); err != nil {
```
Comment (anastasiamac) [inline state/open.go]: Thank you!!!

File: state/open.go (hunk 2 above)
Comment (axw) [inline state/open.go]: It cannot be nil, it's a struct type.

Comment (anastasiamac) [inline state/open.go]: Oh I see. I've read a ghosted asterix :)

**Comment by axw** [comment]
$$merge$$

**Comment by axw** [comment]
$$merge$$

**Pull request merged**